[Merged by Bors] - feat: Linear upper or lower sets topologies are completely normal

Komyyy opened the pull request:
The primary motivation for this theorem is to prove that Sierpiński space is completely normal.

---

- [x] depends on: #28792
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Komyyy committed:
feat: Linear upper or lower sets topologies are completely normal

Komyyy committed:
mk_all

Komyyy committed:
- simp

Komyyy committed:
Merge branch 'master' into linear_upperSetTopology

Komyyy committed:
priority := low & Prop

Komyyy committed:
Merge branch 'master' into linear_upperSetTopology

Komyyy committed:
chore: suggestions

Komyyy committed:
feat: prove for lower sets topologies

alreadydone commented:
Thanks 🎉 
maintainer merge

bryangingechen commented on Mathlib/Topology/Separation/LinearUpperLowerSetTopology.lean:
```diff
@@ -0,0 +1,39 @@
+    obtain (rfl | ⟨b, hb⟩) := t.eq_empty_or_nonempty
+    case inl => simp
+    exfalso
+    -- issue #27058
+    fail_if_success grw [← singleton_subset_iff.mpr ha, ← singleton_subset_iff.mpr hb] at hcst hsct
```
I'm assuming you checked this with current `master` (which includes the PR mentioned at #27058) and it still doesn't work? Could you post an update on that issue, mentioning this PR?

bryangingechen commented:
Thanks!
bors d+

Komyyy committed:
Merge branch 'master' into linear_upperSetTopology

Komyyy committed:
refactor: register the `Disjoint` lemma for `Set`s as a `gcongr` lemma
---
This PR enables `grewrite` to rewrite `Disjoint` goals for `Set`s:
```lean
import Mathlib.Topology.Closure

example {α : Type*} [TopologicalSpace α] {s t u : Set α}
    (hst : s ⊆ t) (htu : Disjoint (closure t) u) : Disjoint (closure s) u := by
  grewrite [← hst] at htu; assumption
```
resolves #27058

Komyyy committed:
Merge branch 'Set.disjoint_of_subset' into linear_upperSetTopology

Komyyy committed:
chore: merge

Komyyy committed:
Merge branch 'master' into linear_upperSetTopology
